### ISSUE TITLE
twisted.python.log.FileLogObserver no longer outputs timestamps

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7711]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)| @wsanchez reported|
|-|-|
|Trac ID|trac#7711|
|Type|enhancement|
|Created|2014-11-11 01:06:51Z|
|Branch|https://github.com/twisted/twisted/tree/log-timestamps-7711|

As noted by ralphm in [#7596](https://github.com/twisted/twisted/issues/7596), twisted.python.log.FileLogObserver no longer outputs timestamps.

Attachments:

* [py3_errors.html](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/09e/09e40ee77d4965c9ddd2edfc8477d1f2b4f5a228/80bf6686556096efb4556312293c1b126d255926.html) (2498329 bytes) - added by wsanchez on 2014-11-14 01:13:30Z - Py3 Trial Log

<details><summary>Searchable metadata</summary>

```
trac-id__7711 7711
type__enhancement enhancement
reporter__wsanchez wsanchez
priority__high high
milestone__New_Logging_System New Logging System
branch__branches_log_timestamps_7711 branches/log-timestamps-7711
branch_author__wsanchez wsanchez
status__closed closed
resolution__fixed fixed
component__logger logger
keywords__ 
time__1415668011680650 1415668011680650
changetime__1481166218030906 1481166218030906
version__None None
owner__wsanchez wsanchez
cc__ralphm cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_1"></a>|@wsanchez commented|
|-|-|

Fixes on branch for [#7596](https://github.com/twisted/twisted/issues/7596):

* r43054 (copy old `time` to new `log_time`)
* r43088, r43090 and r43095 (tests for timestamp in output in old logging system)
* r43091 (LegacyLogObserverWrapper should provide `time` and `system`)
* r43092 (restores timestamps)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_3"></a>|@wsanchez commented|
|-|-|

Waiting on [#7710](https://github.com/twisted/twisted/issues/7710) to avoid conflicts.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_4"></a>|@wsanchez commented|
|-|-|

Merged the revisions noted in comment:1 above in r43464.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_5"></a>|@wsanchez commented|
|-|-|

(In [43470]) Branch for [#7711](https://github.com/twisted/twisted/issues/7711)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez commented|
|-|-|

Reverted r43464 in r43468, as this is what brought in the Py3 failures.

Branched for this, and re-committed on [branches/log-timestamps-7711](branches/log-timestamps-7711) in r43473.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

Isolating the py3 failures… the changes to `_legacy.py` and `test_legacy.py` pass both py2 and py3 tests.

The changes to `test_log.py` break the same three tests ing py2 and py3, which is expected.  So all the py3 brokenness is apparently in `log.py`.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

OK, so, to recap how we got here:

1. @glyph reviewed the changed on this branch and it was good except py3 broke
2. The differences in string handling between py2 and py3 is a pain in the ass.

r43486 addresses the impact of those differences in the following manner:

1. Notice that we're seemingly a bit inconsistent in our string handling in log.py.
  1. `textFromEventDict` returns a //native// string.
  2. `_safeFormat` returns UTF-8 //bytes//.
2. Accept that messages in log.py are native strings. Implications:
  1. `_safeFormat` needs to return a native string.
  2. Tests need minor tweaks.

Note that only the legacy logging uses native strings.  The new logging explicitly uses and expects unicode strings everywhere.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_9"></a>|@wsanchez commented|
|-|-|

To review.

Note the weird process we are using here (see glyph's comment 17 in [#7548](https://github.com/twisted/twisted/issues/7548)), which is treating [branches/log-booyah-6750-5](branches/log-booyah-6750-5) as trunk for the logging work.

Buildbot doesn't seem to care for the branch-from branch thing, so I'll note that I did run the py2 and py3 tests on my mac and it was failing and is now passing.  I'll kick the full buildbot fleet again as part of merging to [branches/log-booyah-6750-5](branches/log-booyah-6750-5) once this is reviewed.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez removed owner<br>@wsanchez set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

So github diff view should be

https://github.com/twisted/twisted/compare/log-booyah-6750-5...log-timestamps-7711

-------

For changes in `_safeFormat` ... since its docstring says that it swallows all errors to always return a string, maybe you should also `ignore` or `replace` for UTF-8 encode/decode errors...  I see that this problem was also present in previous code... so maybe it was not designed to swallow all errors.


--------

``` util.untilConcludes(self.flush)  # Hoorj! ```

Is this comment really useful? It does not say anything to me, other than WTF!?!

-------

Please not that this is not an exhaustive review... but I tried to add my comments :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_12"></a>|@wsanchez commented|
|-|-|

adiroban: that diff looks right for the whole branch.  The part that wasn't previously review by glyph in [#7596](https://github.com/twisted/twisted/issues/7596) is r43486.

`#Hoorj!` is not a great comment.  Nor is it mine… I'm just reverting that code to what it is on trunk.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @wsanchez|
|-|-|

Thanks for keeping with the slog here, wsanchez.  Sorry for the delay!

This change looks pretty good to me, except:

  1. The comment in `_safeFormat` should be a `@return` epytext markup.  Native strings are the worst type, because their semantics change the most between versions, so it's important to be as clear as possible that it's specifically intended that it be native.  And while it's not strictly necessary for this change to be valid, I think it would be good to have `@param`s on `fmtString` and `fmtDict` too, to be clear on what types are supported.
  2. [there are some twistedchecker errors](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2437/steps/run-twistedchecker/logs/new%20twistedchecker%20errors) - some look actually relevant, although there's something wrong with the ratchet builder right now where it keeps reporting stuff in serialport and gtk2reactor so ignore that part obviously.

I noticed that there are some [pydoctor errors](https://buildbot.twistedmatrix.com/builders/documentation/builds/5047/steps/api-documentation/logs/new%20pydoctor%20errors) in the new logger stuff too, although they aren't from this branch; so please address the above points and land on the appropriate branch.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_14"></a>|@wsanchez commented|
|-|-|

[#11338](https://github.com/twisted/twisted/issues/11338) addressed by r43687.

#2… the twistedchecker errors are also not on this branch except one, which is erroneous output; "2 blank lines after docstring" for a callable with no code.

I'll make sure the main branch is clean of twistedchecker and pydoctor noise before it gets merged.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez set status to `closed`|
|-|-|

(In [43688]) Merge branches/log-timestamps-7711: twisted.python.log.FileLogObserver
no longer outputs timestamps

Author: wsanchez
Reviewer: glyph
Fixes: [#7711](https://github.com/twisted/twisted/issues/7711)

Legacy logging lacked some tests to verify that the output is correct
and includes a timestamp, and that broke when integrating the two
systems.  Tests have been added and event dict keys now bridge data
between the two correctly.

Legacy logging also revolves around native strings and attempts to
make it use bytes consistantly result in compatibility issues, so we
are now explicit about our use of native strings.
